### PR TITLE
Fix test groups order in the report

### PR DESCRIPTION
**Pull request by mgalka**
This change adds "index" field to the TestGroup document. The field is supposed to be used for test sub-groups ordering.
For LAVA test results the index field stores the subsequent integer values for each sub-group in a single test run.
This commit adds also fixes sub-groups sorting in the e-mail test report. It uses the 'index' field.

The aim of the provided changes is to avoid e-mail report presenting failed test cases in a different order that they were run by the test backend.

The changes should be enough for LAVA, however they may occur insufficient for test backends that run test groups asynchronously.

Failed tests order in the report before fix:
```
   Control-ioctls-Input-2 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)   

    Control-ioctls-Input-3 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)   

    Control-ioctls-Input-0 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)   

    Control-ioctls-Input-1 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)                         

    Streaming-ioctls_Test-input-0 - 7 tests: 4  PASS, 3 FAIL, 0 SKIP
      * MMAP-epoll: never passed
      * USERPTR-no-poll: failing since 5 days (last pass: kernelci-local-snapshot-032-4-g0227673db427, first fail: kernelci-local-snapshot-032-5-g991fb479edd6)
      * USERPTR-select: failing since 5 days (last pass: kernelci-local-snapshot-032-4-g0227673db427, first fail: kernelci-local-snapshot-032-5-g991fb479edd6)         
```
Failed tests order in the report after fix:
```
   Control-ioctls-Input-0 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)         

    Control-ioctls-Input-1 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)         

    Control-ioctls-Input-2 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)         

    Control-ioctls-Input-3 - 6 tests: 4  PASS, 1 FAIL, 1 SKIP
      * VIDIOC_QUERYCTRL: new failure (last pass: kernelci-local-snapshot-032-5-ge372da4a7adc)         

    Streaming-ioctls_Test-input-0 - 7 tests: 4  PASS, 3 FAIL, 0 SKIP
      * MMAP-epoll: never passed
      * USERPTR-no-poll: failing since 5 days (last pass: kernelci-local-snapshot-032-4-g0227673db427, first fail: kernelci-local-snapshot-032-5-g991fb479edd6)
      * USERPTR-select: failing since 5 days (last pass: kernelci-local-snapshot-032-4-g0227673db427, first fail: kernelci-local-snapshot-032-5-g991fb479edd6)
```

test_group documents with index added in MongoDB
```
/* 5 */
{
    "_id" : ObjectId("5ccad80532f0c904c8951ff5"),
    "index" : 5,
    "name" : "Control-ioctls-Input-0"
}

/* 9 */
{
    "_id" : ObjectId("5ccad80532f0c904c8951ffc"),
    "index" : 9,
    "name" : "Control-ioctls-Input-1"
}

/* 13 */
{
    "_id" : ObjectId("5ccad80532f0c904c8951fe7"),
    "index" : 13,
    "name" : "Control-ioctls-Input-2"
}

/* 17 */
{
    "_id" : ObjectId("5ccad80532f0c904c8951fee"),
    "index" : 17,
    "name" : "Control-ioctls-Input-3"
}

/* 21 */
{
    "_id" : ObjectId("5ccad80532f0c904c8952046"),
    "index" : 21,
    "name" : "Streaming-ioctls_Test-input-0"
}
```

### CODE REVIEW

**Comment by gctucker** [comment]
> test_group documents with index added in MongoDB
> 
> ```
> /* 5 */
> {
>     "_id" : ObjectId("5ccad80532f0c904c8951ff5"),
>     "index" : 5,
>     "name" : "Control-ioctls-Input-0"
> }
> 
> /* 9 */
> {
>     "_id" : ObjectId("5ccad80532f0c904c8951ffc"),
>     "index" : 9,
>     "name" : "Control-ioctls-Input-1"
> }
> 
> /* 13 */
> {
>     "_id" : ObjectId("5ccad80532f0c904c8951fe7"),
>     "index" : 13,
>     "name" : "Control-ioctls-Input-2"
> }
> 
> /* 17 */
> {
>     "_id" : ObjectId("5ccad80532f0c904c8951fee"),
>     "index" : 17,
>     "name" : "Control-ioctls-Input-3"
> }
> 
> /* 21 */
> {
>     "_id" : ObjectId("5ccad80532f0c904c8952046"),
>     "index" : 21,
>     "name" : "Streaming-ioctls_Test-input-0"
> }
> ```

@mgalka Was it expected that the numbers would increase in steps of 4, rather than 1?

File: app/utils/callback/lava.py [hunk 1]
```diff
@@ -43,8 +44,7 @@
 
 TEST_CASE_MAP = {
     models.NAME_KEY: "name",
-    models.STATUS_KEY: "result",
-    models.INDEX_KEY: "logged",
+    models.STATUS_KEY: "result"
```
Comment (gctucker) [inline app/utils/callback/lava.py]: It's better to keep the trailing comma, to reduce the diff noise.

File: app/utils/callback/lava.py [hunk 2]
```diff
@@ -547,10 +549,8 @@ def add_tests(job_data, lab_name, db_options, base_path=utils.BASE_PATH):
     errors = {}
     ex = None
     msg = None
-
```
Comment (gctucker) [inline app/utils/callback/lava.py]: Probably worth leaving these blank lines unless they're really causing an issue, as white space noise causes unnecessary merge conflicts.

**Comment by gctucker** [comment]
Actually the indexes seem OK in the database: 1, 2, 3...
```
> db.test_group.find({kernel: 'v5.1-8123-g396ee6760888', name: 'v4l2-compliance-uvc'}, {name: 1, sub_groups: 1, device_type: 1})[0]
{
	"_id" : ObjectId("5cd3f799cf3a0fe973b2de5c"),
	"name" : "v4l2-compliance-uvc",
	"device_type" : "rk3288-veyron-jaq",
	"sub_groups" : [
		ObjectId("5cd3f799cf3a0fe973b2de5d"),
		ObjectId("5cd3f799cf3a0fe973b2de60"),
		ObjectId("5cd3f799cf3a0fe973b2de65"),
		ObjectId("5cd3f799cf3a0fe973b2de68"),
		ObjectId("5cd3f799cf3a0fe973b2de77"),
		ObjectId("5cd3f799cf3a0fe973b2de7e"),
		ObjectId("5cd3f799cf3a0fe973b2de89"),
		ObjectId("5cd3f799cf3a0fe973b2de8d"),
		ObjectId("5cd3f799cf3a0fe973b2de91")
	]
}
> db.test_group.findOne(ObjectId("5cd3f799cf3a0fe973b2de5d"), {name: 1, index: 1})
{
	"_id" : ObjectId("5cd3f799cf3a0fe973b2de5d"),
	"name" : "Required-ioctls",
	"index" : 1
}
> db.test_group.findOne(ObjectId("5cd3f799cf3a0fe973b2de60"), {name: 1, index: 1})
{
	"_id" : ObjectId("5cd3f799cf3a0fe973b2de60"),
	"name" : "Allow-for-multiple-opens",
	"index" : 2
}
> db.test_group.findOne(ObjectId("5cd3f799cf3a0fe973b2de65"), {name: 1, index: 1})
{
	"_id" : ObjectId("5cd3f799cf3a0fe973b2de65"),
	"name" : "Debug-ioctls",
	"index" : 3
}
> db.test_group.findOne(ObjectId("5cd3f799cf3a0fe973b2de68"), {name: 1, index: 1})
{
	"_id" : ObjectId("5cd3f799cf3a0fe973b2de68"),
	"name" : "Input-ioctls",
	"index" : 4
}
> db.test_group.findOne(ObjectId("5cd3f799cf3a0fe973b2de77"), {name: 1, index: 1})
{
	"_id" : ObjectId("5cd3f799cf3a0fe973b2de77"),
	"name" : "Control-ioctls-Input-0",
	"index" : 5
}
> db.test_group.findOne(ObjectId("5cd3f799cf3a0fe973b2de7e"), {name: 1, index: 1})
{
	"_id" : ObjectId("5cd3f799cf3a0fe973b2de7e"),
	"name" : "Format-ioctls-Input-0",
	"index" : 6
}
```

**Comment by gctucker** [comment]
For the record, today's email from staging which seems to be showing all the results in the correct order:
```
kernelci/staging.kernelci.org v4l2-compliance on uvcvideo: 104 tests, 0 regressions (v5.1-8123-g396ee6760888)

Test results summary
--------------------

V4L2 Compliance on the uvcvideo driver.

This test ran "v4l2-compliance -s" from v4l-utils:

    https://www.linuxtv.org/wiki/index.php/V4l2-utils

See each detailed section in the report below to find out the git URL and
particular revision that was used to build the test binaries.


  Tree:    kernelci
  Branch:  staging.kernelci.org
  Kernel:  v5.1-8123-g396ee6760888
  URL:     https://github.com/kernelci/linux.git
  Commit:  396ee6760888509476cb18d56a53df223b1b28e6


1  | rk3288-veyron-jaq      | arm   |  52 total:  17 PASS   9 FAIL  26 SKIP
2  | rk3399-gru-kevin       | arm64 |  52 total:  17 PASS   9 FAIL  26 SKIP  
    

Test failures
------------- 
1  | rk3288-veyron-jaq      | arm   |  52 total:  17 PASS   9 FAIL  26 SKIP

  Config:      multi_v7_defconfig
  Compiler:    gcc-8 (arm-linux-gnueabihf-gcc (Debian 8.3.0-2) 8.3.0)
  Lab Name:    lab-collabora-dev
  Plain log:   http://staging-storage.kernelci.org/kernelci/staging.kernelci.org/v5.1-8123-g396ee6760888/arm/multi_v7_defconfig/gcc-8/lab-collabora-dev/v4l2-compliance-uvc-rk3288-veyron-jaq.txt
  HTML log:    http://staging-storage.kernelci.org/kernelci/staging.kernelci.org/v5.1-8123-g396ee6760888/arm/multi_v7_defconfig/gcc-8/lab-collabora-dev/v4l2-compliance-uvc-rk3288-veyron-jaq.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch-v4l2/20190506.0/armhf/rootfs.cpio.gz
  Test Git:    git://linuxtv.org/v4l-utils.git
  Test Commit: 5b59b0fd7a5d37921c64e8640527e1180e3e289a
              

    Control-ioctls-Input-0 - 6 tests: 3  PASS, 2 FAIL, 1 SKIP
      * VIDIOC_QUERY_EXT_CTRL/QUERYMENU: never passed
      * VIDIOC_G/S/TRY_EXT_CTRLS: never passed   

    Format-ioctls-Input-0 - 10 tests: 4  PASS, 1 FAIL, 5 SKIP
      * VIDIOC_G/S_PARM: never passed     

    Buffer-ioctls-Input-0 - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
      * Requests: never passed   

    Streaming-ioctls_Test-input-0 - 7 tests: 1  PASS, 5 FAIL, 1 SKIP
      * MMAP-no-poll: never passed
      * MMAP-select: never passed
      * MMAP-epoll: never passed
      * USERPTR-no-poll: never passed
      * USERPTR-select: never passed   
  
    

Test failures
------------- 
2  | rk3399-gru-kevin       | arm64 |  52 total:  17 PASS   9 FAIL  26 SKIP

  Config:      defconfig
  Compiler:    clang-8 (clang version 8.0.1-svn359952-1~exp1~20190504004501.65 (branches/release_80))
  Lab Name:    lab-collabora-dev
  Plain log:   http://staging-storage.kernelci.org/kernelci/staging.kernelci.org/v5.1-8123-g396ee6760888/arm64/defconfig/clang-8/lab-collabora-dev/v4l2-compliance-uvc-rk3399-gru-kevin.txt
  HTML log:    http://staging-storage.kernelci.org/kernelci/staging.kernelci.org/v5.1-8123-g396ee6760888/arm64/defconfig/clang-8/lab-collabora-dev/v4l2-compliance-uvc-rk3399-gru-kevin.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch-v4l2/20190506.0/arm64/rootfs.cpio.gz
  Test Git:    git://linuxtv.org/v4l-utils.git
  Test Commit: 5b59b0fd7a5d37921c64e8640527e1180e3e289a
              

    Control-ioctls-Input-0 - 6 tests: 3  PASS, 2 FAIL, 1 SKIP
      * VIDIOC_QUERY_EXT_CTRL/QUERYMENU: never passed
      * VIDIOC_G/S/TRY_EXT_CTRLS: never passed   

    Format-ioctls-Input-0 - 10 tests: 4  PASS, 1 FAIL, 5 SKIP
      * VIDIOC_G/S_PARM: never passed     

    Buffer-ioctls-Input-0 - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
      * Requests: never passed   

    Streaming-ioctls_Test-input-0 - 7 tests: 1  PASS, 5 FAIL, 1 SKIP
      * MMAP-no-poll: never passed
      * MMAP-select: never passed
      * MMAP-epoll: never passed
      * USERPTR-no-poll: never passed
      * USERPTR-select: never passed   
 ```

**Comment by mgalka** [comment]
> @mgalka Was it expected that the numbers would increase in steps of 4, rather than 1?
The index increases in steps of 1, the documents listed are only the ones that failed.

**Comment by gctucker** [comment]
> > @mgalka Was it expected that the numbers would increase in steps of 4, rather than 1?
> 
> The index increases in steps of 1, the documents listed are only the ones that failed.

OK, thanks.  Also I checked in the staging db directly and it's all fine (see my previous comment).

File: app/utils/callback/lava.py (hunk 1 above)
Comment (mgalka) [reply app/utils/callback/lava.py]: fixed

File: app/utils/callback/lava.py (hunk 2 above)
Comment (mgalka) [reply app/utils/callback/lava.py]: fixed

**Comment by gctucker** [comment]
Tested again today, same results after the clean-up so this is all good to go :)

**Pull request merged**